File: src/main/java/org/example/util/Batch.java
package org.example.util;

import java.time.Instant;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.TimeUnit;

public class Batch {

    static final int MAX = 1;
    private static final ScheduledExecutorService scheduler = Executors.newScheduledThreadPool(1);

    private Batch() {
        throw new IllegalStateException("Utility class");
    }

    /**
     * @param minutes minutes to sleep until processing
     */
    public synchronized static void process(final int minutes, final Runnable runnable) {
        scheduler.scheduleAtFixedRate(() -> {
            try {
                System.out.println("Starting batch processing... at " + Instant.now());
                runnable.run();
            } catch (Exception e) {
                System.err.println("Error processing batch at: " + Instant.now() + " - " + e.getMessage());
            }
        }, minutes, minutes, TimeUnit.MINUTES);
    }
}

File: src/main/java/org/example/util/Batch.kt
package org.example.util

import java.time.Instant
import java.util.concurrent.Executors
import java.util.concurrent.ScheduledExecutorService
import java.util.concurrent.TimeUnit

class Batch private constructor() {
    init {
        throw IllegalStateException("Utility class")
    }

    companion object {
        const val MAX: Int = 1
        private val scheduler: ScheduledExecutorService = Executors.newScheduledThreadPool(1)

        /**
         * @param minutes minutes to sleep until processing
         */
        @JvmStatic
        @Synchronized
        fun process(minutes: Int, runnable: Runnable) {
            scheduler.scheduleAtFixedRate(Runnable {
                try {
                    println("Starting batch processing... at " + Instant.now())
                    runnable.run()
                } catch (e: Exception) {
                    System.err.println("Error processing batch at: " + Instant.now() + " - " + e.message)
                }
            }, minutes.toLong(), minutes.toLong(), TimeUnit.MINUTES)
        }
    }
}